### PR TITLE
Reapply [lldb][DWARF] Delay struct/class/union definition DIE searching when parsing declaration DIEs.

### PR DESCRIPTION
**Pull request by ZequanWu**
This is a reapply of https://github.com/llvm/llvm-project/pull/92328 and https://github.com/llvm/llvm-project/pull/93839.

It now passes the [test](https://github.com/llvm/llvm-project/commit/de3f1b6d68ab8a0e827db84b328803857a4f60df), which crashes with the original reverted changes.

### CODE REVIEW

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 1]
```diff
@@ -1603,41 +1603,74 @@ DWARFASTParserClang::GetCPlusPlusQualifiedName(const DWARFDIE &die) {
...
+        // entry in the map is a declaration, then we need to update the entry
+        // to point to the definition DIE.
+        if (!attrs.is_forward_declaration &&
+            unique_ast_entry_type->m_is_forward_declaration) {
+          unique_ast_entry_type->m_die = die;
+          unique_ast_entry_type->m_byte_size = byte_size;
+          unique_ast_entry_type->m_declaration = unique_decl;
+          unique_ast_entry_type->m_is_forward_declaration = false;
+          // Need to update Type ID to refer to the definition DIE. because
+          // it's used in ParseSubroutine to determine if we need to copy cxx
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:1661]: The code in question has moved to ParseCXXMethod in the mean time.

File: lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp [hunk 2]
```diff
@@ -1631,26 +1638,48 @@ bool SymbolFileDWARF::CompleteType(CompilerType &compiler_type) {
...
+    }
+  }
+  if (!def_die) {
+    // No definition found. Proceed with the declaration die. We can use it to
+    // create a forward-declared type.
+    def_die = decl_die;
+  }
 
-    Type *type = GetDIEToType().lookup(dwarf_die.GetDIE());
+  Type *type = ResolveType(def_die);
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:1661]: This still creates the lookup loop where we use the type to find the declaration DIE, use the decl DIE to find the definition DIE, and then use that DIE to find the type again. I really wanted to avoid that, for several reasons:
- it's fragile, in that if this call returns a type different from the one we started, we will end up in an inconsistent state. I think this doesn't crash right now because my change to CompleteTypeFromDWARF has made it more robust to these kinds of inconsistent inputs. However, that wasn't my goal, and I don't think we should be relying on that.
-  it's doing more work than it should. We already *know* which type we're supposed to be completing. Why are we looking it up again?
- it makes the code hard to follow

I was hoping this code would basically boil down to something like (with additional error checks, probably):
```
dwarf_ast->CompleteTypeFromDWARF(FindDefinitionDIE(decl_die), GetDIEToType().lookup(decl_die.GetDIE()), compiler_type);
```

Why do we need to do all of the other stuff? Can we refactor something so that we don't need to do all those other things here?

**Review by labath — COMMENTED** [review]
I didn't review the whole patch yet because I think we first need to discuss the type lookup question (see my big inline comment).

File: lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp (hunk 2 above)
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:1661]: `ResolveType(def_die)` just do [this part](https://github.com/llvm/llvm-project/pull/98361/files/37b6878b9125c314c75053f7d5b0ba520111e9a3#diff-5dd6736e4d6c38623630df16c4494c1a8b099716ee0f05c9af54b4bafb1d864eR1607-R1674) of work and then return the type, which was created from the declaration DIE. It doesn't create a loop or recursion. We need it to do the work when finding existing type in the UniqueDWARFASTTypeMap, which is basically this block: https://github.com/llvm/llvm-project/pull/98361/files/37b6878b9125c314c75053f7d5b0ba520111e9a3#diff-5dd6736e4d6c38623630df16c4494c1a8b099716ee0f05c9af54b4bafb1d864eR1648-R1674. Those code are important to keep some maps(`GetDIEToType()`, `LinkDeclContextToDIE`, `UniqueDWARFASTTypeMap`) updated.

Comment (labath) [reply lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:1661]: > We need it to do the work when finding existing type in the UniqueDWARFASTTypeMap,

Why do we need to find that type? We already know which type we're completing (we get it as an argument to `SymbolFileDWARF::CompleteType`)...

> It doesn't create a loop or recursion. We need it to do the work when finding existing type in the UniqueDWARFASTTypeMap, which is basically this block: https://github.com/llvm/llvm-project/pull/98361/files/37b6878b9125c314c75053f7d5b0ba520111e9a3#diff-5dd6736e4d6c38623630df16c4494c1a8b099716ee0f05c9af54b4bafb1d864eR1648-R1674.

That's only true if `UniqueDWARFASTTypeMap` functions perfectly and returns the same type we started out with. My point is we shouldn't/don't need to rely on that.

> Those code are important to keep some maps(`GetDIEToType()`, `LinkDeclContextToDIE`, `UniqueDWARFASTTypeMap`) updated

Yes, we need to update those maps, but I think we can/should just update those directly, instead of relying on it to be populated as a side-effect of the second ParseTypeFromDWARF call.  Like, we already have the type and the definition die, so why don't we just do a `GetDIEToType()[defn_die] = type`? And for the UniqueDWARFASTTypeMap, we can use the original declaration die to perform the lookup. This will be both faster (we can add a specialized lookup function that assumes that the lookup die matches one of the `udt.m_die` members) and guaranteed to be correct.

Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp]: That makes sense. I created another function to update those maps and use decl_die to lookup in UniqueDWARFASTTypeMap.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 3]
```diff
@@ -824,6 +824,36 @@ DWARFASTParserClang::GetDIEClassTemplateParams(const DWARFDIE &die) {
...
+  ConstString unique_typename(decl_attrs.name);
+  Declaration decl_declaration(decl_attrs.decl);
+  if (Language::LanguageIsCPlusPlus(
+          SymbolFileDWARF::GetLanguage(*decl_die.GetCU()))) {
+    std::string qualified_name = GetCPlusPlusQualifiedName(decl_die);
+    if (!qualified_name.empty())
+      unique_typename = ConstString(qualified_name);
+    decl_declaration.Clear();
+  }
+  if (UniqueDWARFASTType *unique_ast_entry_type =
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:843]: This should always be true, right (because we've already parsed the type and inserted the proper entry? Failing to find the entry here probably deserves at least a log message..

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 4]
```diff
@@ -1603,41 +1633,76 @@ DWARFASTParserClang::GetCPlusPlusQualifiedName(const DWARFDIE &die) {
...
+          unique_ast_entry_type->m_die = die;
+          if (byte_size)
+            unique_ast_entry_type->m_byte_size = byte_size;
+          if (unique_decl.IsValid())
+            unique_ast_entry_type->m_declaration = unique_decl;
+          unique_ast_entry_type->m_is_forward_declaration = false;
+          // Need to update Type ID to refer to the definition DIE. because
+          // it's used in ParseCXXMethod to determine if we need to copy cxx
+          // method types from a declaration DIE to this definition DIE.
+          type_sp->SetID(die.GetID());
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:1695]: And this could be a helper function to deduplicate the second part of `MappingDeclDIEToDefDIE`

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 5]
```diff
@@ -824,6 +824,36 @@ DWARFASTParserClang::GetDIEClassTemplateParams(const DWARFDIE &die) {
   return {};
 }
 
+void DWARFASTParserClang::MappingDeclDIEToDefDIE(
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:827]: nit: Maybe you're reading this another way, but this name does not sound like a verb phrase to me (which I think it should be, as the function changes the state of the program). I'd just call it `MapDeclDIEToDefDIE`

File: lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp [hunk 6]
```diff
@@ -1631,27 +1638,49 @@ bool SymbolFileDWARF::CompleteType(CompilerType &compiler_type) {
...
+    return false;
+  Type *type = GetDIEToType().lookup(decl_die.GetDIE());
+  if (decl_die != def_die) {
+    GetDIEToType()[def_die.GetDIE()] = type;
+    // Need to update Type ID to refer to the definition DIE. because
+    // it's used in ParseCXXMethod to determine if we need to copy cxx
+    // method types from a declaration DIE to this definition DIE.
+    type->SetID(def_die.GetID());
+    if (DWARFASTParserClang *ast_parser =
+            static_cast<DWARFASTParserClang *>(dwarf_ast))
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp:1672]: this check is superfluous. static_cast will never turn a non-null pointer into nullptr.

File: lldb/test/Shell/SymbolFile/DWARF/x86/simple-template-names-context.cpp [hunk 7]
```diff
@@ -4,8 +4,8 @@
 
 // REQUIRES: lld
 
-// RUN: %clang --target=x86_64-pc-linux -c %s -o %t-a.o -g -gsimple-template-names -DFILE_A
-// RUN: %clang --target=x86_64-pc-linux -c %s -o %t-b.o -g -gsimple-template-names -DFILE_B
```
Comment (labath) [inline lldb/test/Shell/SymbolFile/DWARF/x86/simple-template-names-context.cpp:8]: We should keep the existing test, but you can add another set of RUN lines to it.
If you do that, it would also be good to rename the test to reflect its wider scope. Maybe to something about type definition searching, as that's the common theme in the two cases.

**Review by labath — COMMENTED** [review]
Thanks. I think this looks pretty good now, just a couple of details to sort out.

File: lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp (hunk 6 above)
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/SymbolFileDWARF.cpp]: Removed.

File: lldb/test/Shell/SymbolFile/DWARF/x86/simple-template-names-context.cpp (hunk 7 above)
Comment (ZequanWu) [reply lldb/test/Shell/SymbolFile/DWARF/x86/simple-template-names-context.cpp]: Added another set of RUN for -fdebug-types-section

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp (hunk 5 above)
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp]: Done.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 8]
```diff
@@ -824,6 +824,36 @@ DWARFASTParserClang::GetDIEClassTemplateParams(const DWARFDIE &die) {
...
+  ParsedDWARFTypeAttributes def_attrs(def_die);
+  ConstString unique_typename(decl_attrs.name);
+  Declaration decl_declaration(decl_attrs.decl);
+  if (Language::LanguageIsCPlusPlus(
+          SymbolFileDWARF::GetLanguage(*decl_die.GetCU()))) {
+    std::string qualified_name = GetCPlusPlusQualifiedName(decl_die);
+    if (!qualified_name.empty())
+      unique_typename = ConstString(qualified_name);
+    decl_declaration.Clear();
+  }
```
[the remark this answers is not in the document]
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:842]: I don't see any equivalent code in here and ParseTypeFromDWARF.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp (hunk 3 above)
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:843]: Yes, this should always be true. Added a log.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp (hunk 4 above)
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp]: Done.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (labath) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:842]: Sorry, I meant `ParseStructureLikeDIE` (which is called from ParseTypeFromDWARF), roughly lines 1664--1673

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 9]
```diff
@@ -1595,49 +1627,67 @@ DWARFASTParserClang::GetCPlusPlusQualifiedName(const DWARFDIE &die) {
...
+          unique_ast_entry_type->UpdateToDefDIE(die, unique_decl, byte_size);
+          clang_type = type_sp->GetForwardCompilerType();
+
+          CompilerType compiler_type_no_qualifiers =
+              ClangUtil::RemoveFastQualifiers(clang_type);
+          auto result = dwarf->GetForwardDeclCompilerTypeToDIE().try_emplace(
+              compiler_type_no_qualifiers.GetOpaqueQualType(),
+              *die.GetDIERef());
+          if (!result.second)
+            result.first->second = *die.GetDIERef();
```
Comment (labath) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:1689]: All this is just `dwarf->GetForwardDeclCompilerTypeToDIE().insert_or_assign(compiler_type_no_qualifiers.GetOpaqueQualType(), *die.GetDIERef())`, right?

**Review by labath — APPROVED** [review]
Looks good. Third time's the charm.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 10]
```diff
@@ -1595,49 +1627,67 @@ DWARFASTParserClang::GetCPlusPlusQualifiedName(const DWARFDIE &die) {
   if (qualified_name.empty())
     qualified_name.append("::");
 
-  qualified_name.append(name);
+  qualified_name.append(unique_typename.GetCString());
   qualified_name.append(GetDIEClassTemplateParams(die));
 
-  return qualified_name;
+  unique_typename = ConstString(qualified_name);
```
Comment (Michael137) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:1633]: Nit: I'd find it slightly clearer if we returned the new name like the old function did (probably don't even need the parameter to be a `ConstString`)

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp [hunk 11]
```diff
@@ -1659,128 +1709,56 @@ DWARFASTParserClang::ParseStructureLikeDIE(const SymbolContext &sc,
...
+            static_cast<void *>(this), die.GetID(), DW_TAG_value_to_name(tag),
+            tag, attrs.name.GetCString(), type_sp->GetID());
       }
+      return type_sp;
     }
   }
 
-  DWARFDIE def_die;
   if (attrs.is_forward_declaration) {
-    Progress progress(llvm::formatv(
```
Comment (Michael137) [inline lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:1717]: Can we move this into `FindDefinitionDIE`? We previously put this here to keep track of when we call into `FindDefinitionDIE` during DWARF parsing

**Review by Michael137 — APPROVED** [review]
I like where we ended up with this. LGTM (with two minor nits)

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp (hunk 11 above)
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:1717]: I moved it inside `FindDefinitionDIE` and changed the logging message to `Searching definition DIE in ...` as that function just do searching not parsing.

File: lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp (hunk 10 above)
Comment (ZequanWu) [reply lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParserClang.cpp:1632]: This functions modifies both `unique_typename` and `decl_declaration` if it's c++, so there're two out values. Returning one out value while modifying another out value via reference parameter looks strange.

**Comment by ZequanWu** [comment]
This buildkite seems got stuck somehow, no logging at all: https://buildkite.com/llvm-project/github-pull-requests/builds/81790#0190bca9-bde7-4fad-8478-9dffd4f669f7. Will merge without waiting for it to finish.
